### PR TITLE
Added Redis-Cache and Sorting functionality

### PR DESCRIPTION
**Pull request by aasif-iqbal**
1. Add Redis-Cache to the quiz controller to enhance performance which reduces execution time by more than half. also added screenshot for ease.

2. Added sorting functionality based on the difficulty level (easy, medium, hard) of quiz questions to enhance the user experience.

### CODE REVIEW

**Comment by aasif-iqbal** [comment]
1. Create a separate route for fetching data redis ie. quiz/redis that doesn't affect application.

2. Add Proper documentation - What is redis cache, working, installation(window, macos & Linux, how it work with nodejs, some common redis commands  in RedisCache.md - (markdown file)

File: backend/.env.example [hunk 1]
```diff
@@ -0,0 +1,12 @@
+PORT=""
```
Comment (sayedazharsabri) [inline backend/.env.example:1]: The .env file is not required in this project

File: backend/.gitignore [hunk 2]
```diff
@@ -0,0 +1,3 @@
+./node_modules
```
Comment (sayedazharsabri) [inline backend/.gitignore:1]: .gitignore is already exist on the project, so not required

File: backend/RedisCache.md [hunk 3]
```diff
@@ -0,0 +1,309 @@
+# Redis Cache
```
Comment (sayedazharsabri) [inline backend/RedisCache.md:1]: Not a redis tutorial

File: backend/package.json [hunk 4]
```diff
@@ -8,12 +8,15 @@
   "dependencies": {
     "@types/bcryptjs": "^2.4.3",
     "@types/cors": "^2.8.16",
+    "@types/dotenv": "^8.2.0",
```
Comment (sayedazharsabri) [inline backend/package.json:11]: not required

File: backend/package.json [hunk 5]
```diff
@@ -8,12 +8,15 @@
...
     "@types/express": "^4.17.17",
     "@types/jsonwebtoken": "^9.0.2",
     "@types/nodemailer": "^6.4.9",
     "@types/otp-generator": "^4.0.1",
+    "@types/redis": "^4.0.11",
```
Comment (sayedazharsabri) [inline backend/package.json:16]: should be in dev dependency

File: backend/package.json [hunk 6]
```diff
@@ -8,12 +8,15 @@
...
     "bcryptjs": "^2.4.3",
     "cors": "^2.8.5",
+    "dotenv": "^16.4.5",
```
Comment (sayedazharsabri) [inline backend/package.json:19]: not required

File: backend/screenshots/Redis_Diagram _light.jpg
Comment (sayedazharsabri) [inline backend/screenshots/Redis_Diagram _light.jpg:1]: Not a redis tutorial

File: backend/screenshots/Redis_Diagram_dark.jpg
Comment (sayedazharsabri) [inline backend/screenshots/Redis_Diagram_dark.jpg:1]: not a redis tutorial

File: backend/screenshots/Without-RedisCache.png
Comment (sayedazharsabri) [inline backend/screenshots/Without-RedisCache.png:1]: Can explain in words and it also depend on system

File: backend/screenshots/with-RedisCache.png
Comment (sayedazharsabri) [inline backend/screenshots/with-RedisCache.png:1]: It can change with system

File: backend/src/app.ts [hunk 7]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
     console.log(error);
   }
 })();
+
+// Redis connection
```
Comment (sayedazharsabri) [inline backend/src/app.ts:96]: Improper comment

File: backend/src/app.ts [hunk 8]
```diff
@@ -20,6 +21,9 @@ const connectionString = process.env.CONNECTION_STRING || "";
 
 const port = process.env.PORT;
 
+const REDIS_PORT: number = 6379;
```
Comment (sayedazharsabri) [inline backend/src/app.ts:24]: Dont hardcode values

File: backend/src/app.ts [hunk 9]
```diff
@@ -1,6 +1,7 @@
 import express, { NextFunction, Request, Response } from "express";
 import mongoose from "mongoose";
 import cors from "cors"
+import * as redis from "redis"
```
Comment (sayedazharsabri) [inline backend/src/app.ts:4]: Dont import everything if not using

File: backend/src/app.ts [hunk 10]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
     console.log(error);
   }
 })();
+
+// Redis connection
+(async () => {
+  try {
+    redisClient = redis.createClient();
+
```
Comment (sayedazharsabri) [inline backend/src/app.ts:100]: why unused new line?

File: backend/src/app.ts [hunk 11]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
...
+    redisClient.on("error", (error) => console.error(`Redis Error : ${error}`));
+
+    console.log("Connected to Redis server.");
+
```
Comment (sayedazharsabri) [inline backend/src/app.ts:104]: New line not required

File: backend/src/app.ts [hunk 12]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
...
   }
 })();
+
+// Redis connection
+(async () => {
+  try {
+    redisClient = redis.createClient();
+
+    redisClient.on("error", (error) => console.error(`Redis Error : ${error}`));
+
```
Comment (sayedazharsabri) [inline backend/src/app.ts:102]: New line not required

File: backend/src/app.ts [hunk 13]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
...
+    console.log("Connected to Redis server.");
+
+
```
Comment (sayedazharsabri) [inline backend/src/app.ts:105]: New line not required

File: backend/src/app.ts [hunk 14]
```diff
@@ -88,3 +92,19 @@ clearBlacklistedTokenScheduler;
     console.log(error);
   }
 })();
+
+// Redis connection
+(async () => {
+  try {
+    redisClient = redis.createClient();
+
+    redisClient.on("error", (error) => console.error(`Redis Error : ${error}`));
```
Comment (sayedazharsabri) [inline backend/src/app.ts:101]: Why not using the existing ProjectError

File: backend/src/controllers/auth.ts [hunk 15]
```diff
@@ -8,9 +8,7 @@ import User from "../models/user";
 import sendEmail from "../utils/email";
 import { ReturnResponse } from "../utils/interfaces";
 import Mailgen from 'mailgen';
-
-
-import OTP from "../models/otp"
+import OTP from "../models/OTP";
```
Comment (sayedazharsabri) [inline backend/src/controllers/auth.ts:11]: Why did you change a file name from lower case to upper case, the file name should be in lower case( if it is not a class) or at least it should be consistent in the whole project

File: backend/src/controllers/otp.ts [hunk 16]
```diff
@@ -1,5 +1,5 @@
 import sendEmail from "../utils/email";
-import OTP from "../models/otp"
+import OTP from "../models/OTP";
```
Comment (sayedazharsabri) [inline backend/src/controllers/otp.ts:2]: Always respect consistency in a project, all file names are small so please take care of this

File: backend/src/controllers/quiz.ts [hunk 17]
```diff
@@ -1,12 +1,30 @@
 //model
-import { RequestHandler } from "express";
+import { RequestHandler, json } from "express";
 
 import ProjectError from "../helper/error";
 import Quiz from "../models/quiz";
 import { ReturnResponse } from "../utils/interfaces";
 
+import * as redis from "redis"
```
Comment (sayedazharsabri) [inline backend/src/controllers/quiz.ts:8]: why importing everything?

File: backend/src/controllers/quiz.ts [hunk 18]
```diff
@@ -1,12 +1,30 @@
...
+import User from "../models/user";
+
+const _createRedisClient = async () => {
+  
+  const client = await redis.createClient();
+  
+  client.on('error', (err) => {
```
Comment (sayedazharsabri) [inline backend/src/controllers/quiz.ts:15]: Why not using async await

File: backend/src/controllers/quiz.ts [hunk 19]
```diff
@@ -1,12 +1,30 @@
...
+      console.error('Redis connection error:', err);
+  });
+  
+  return client;
+}
+
 const createQuiz: RequestHandler = async (req, res, next) => {
   try {
+    
+    const redisClient = await _createRedisClient();
```
Comment (sayedazharsabri) [inline backend/src/controllers/quiz.ts:25]: Wrong way of implementation, why creating redis client in each call?

File: backend/src/controllers/quiz.ts (hunk 17 above)
Comment (sayedazharsabri) [inline backend/src/controllers/quiz.ts:8]: Why importing redis on quiz manager page, not good way for the implementaion of cache

File: backend/src/controllers/quiz.ts [hunk 20]
```diff
@@ -18,6 +36,10 @@ const createQuiz: RequestHandler = async (req, res, next) => {
     const isPublicQuiz = req.body.isPublicQuiz;
     const allowedUser = req.body.allowedUser;
     const quiz = new Quiz({ name, category, difficultyLevel, questionList, answers, passingPercentage, createdBy, attemptsAllowedPerUser, isPublicQuiz, allowedUser });
+    
+    //clean up old cache data
+    await redisClient.del(JSON.stringify(createdBy));
```
Comment (sayedazharsabri) [inline backend/src/controllers/quiz.ts:41]: It makes no sense

**Review by sayedazharsabri — COMMENTED** [review]
- The way you have implemented the redis in the project is wrong.
- File renaming was incorrect.
- Adding extra environment files is makes no senses
- Add types as dev depencies
- I will reject the PR because of above changes and not a good way of implementation
